### PR TITLE
feat: Add Customer Search Demo App

### PR DESCRIPTION
**Pull request by CodeChamp-SS**
# Description

Thank you for opening a Pull Request!
Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [ ] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [ ] You are listed as the author in your notebook or README file.
  - [ ] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [ ] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [ ] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

**Review by holtskinner — CHANGES_REQUESTED** [review]
See comment. I will do a more thorough review once the files are cleaned up

**Comment by CodeChamp-SS** [comment]
> @CodeChamp-SS Can you add a description of your sample app to the PR Description?
> 
> Also, all of the Images/PDF/CSV files make this PR quite large. These files should really be added to a Google Cloud Storage Bucket and linked in if needed, to reduce the size of the repository.
> 
> We have a public Google Cloud Storage Bucket [`gs://github-repo/`](https://console.cloud.google.com/storage/browser/github-repo) for related files for Notebooks/Sample Apps in this repo. I can upload files for you.

@holtskinner you may put the following files/folders in the GCS bucket:
gemini/sample-apps/customer-search/files/tables
gemini/sample-apps/customer-search/files/CymbalBankWebDeployed/assets
gemini/sample-apps/customer-search/files/FD_TnC.pdf
gemini/sample-apps/customer-search/files/Market_Summary.pdf

**Comment by holtskinner** [comment]
@CodeChamp-SS I moved the requested files to GCS:

`gs://github-repo/generative-ai/sample-apps/customer-search/`

New Paths:

- `gemini/sample-apps/customer-search/files/tables` → `gs://github-repo/generative-ai/sample-apps/customer-search/tables`
- `gemini/sample-apps/customer-search/files/CymbalBankWebDeployed/assets` → `gs://github-repo/generative-ai/sample-apps/customer-search/assets/`
- `gemini/sample-apps/customer-search/files/FD_TnC.pdf` → `gs://github-repo/generative-ai/sample-apps/customer-search/FD_TnC.pdf`
- `gemini/sample-apps/customer-search/files/Market_Summary.pdf` → `gs://github-repo/generative-ai/sample-apps/customer-search/Market_Summary.pdf`

You can also access these through the public links https://storage.googleapis.com/github-repo/generative-ai/sample-apps/customer-search/FD_TnC.pdf

I also recommend moving the README images to GCS and linking them in this way.

File: gemini/sample-apps/customer-search/functions/credit-card-imagen/main.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        return "A dog reading the newspaper"
+
+
+def generate_base64_image(user_prompt):
+    vertexai.init(project=PROJECT_ID, location=LOCATION)
+    model = ImageGenerationModel.from_pretrained("imagegeneration@005")
+    images = model.generate_images(prompt=user_prompt, number_of_images=1, seed=1)
+
+    try:
+        images[0].save(location="./gen-img1.png", include_generation_parameters=True)
```
Comment (holtskinner) [inline gemini/sample-apps/customer-search/functions/credit-card-imagen/main.py:28]: You should be able to access the image base64 directly. https://github.com/googleapis/python-aiplatform/blob/118826ff88d6e0510dc67e37edfc16f646cb3489/vertexai/vision_models/_vision_models.py#L103C9-L103C21

File: gemini/sample-apps/customer-search/functions/event-recommendationv2/main.py
Comment (holtskinner) [inline gemini/sample-apps/customer-search/functions/event-recommendationv2/main.py:1]: Why is there an eventrecommendation v2 and v3?

**Review by holtskinner — CHANGES_REQUESTED** [review]
I'm stopping part way through to allow time to make some of the changes.

Overall there's a lot of duplicated code and most of this application seems to be overly complex. I would highly recommend refactoring and finding ways to reduce duplication and complexity. It will make the application code much more readable/understandable/maintainable.

**Comment by CodeChamp-SS** [comment]
> @CodeChamp-SS I moved the requested files to GCS:
> 
> `gs://github-repo/generative-ai/sample-apps/customer-search/`
> 
> New Paths:
> 
> * `gemini/sample-apps/customer-search/files/tables` → `gs://github-repo/generative-ai/sample-apps/customer-search/tables`
> * `gemini/sample-apps/customer-search/files/CymbalBankWebDeployed/assets` → `gs://github-repo/generative-ai/sample-apps/customer-search/assets/`
> * `gemini/sample-apps/customer-search/files/FD_TnC.pdf` → `gs://github-repo/generative-ai/sample-apps/customer-search/FD_TnC.pdf`
> * `gemini/sample-apps/customer-search/files/Market_Summary.pdf` → `gs://github-repo/generative-ai/sample-apps/customer-search/Market_Summary.pdf`
> 
> You can also access these through the public links https://storage.googleapis.com/github-repo/generative-ai/sample-apps/customer-search/FD_TnC.pdf
> 
> I also recommend moving the remaining images for the README and app to GCS and linking them in this way.

Thanks! Will replace the images/files in the code with these.

File: gemini/sample-apps/customer-search/files/CymbalBankWebDeployed/img/build-config.png
[the remark this answers is not in the document]
Comment (CodeChamp-SS) [reply gemini/sample-apps/customer-search/files/CymbalBankWebDeployed/img/build-config.png:1]: No they're no needed, I will remove these.

File: gemini/sample-apps/customer-search/files/CymbalBankWebDeployed/public/cymbal-credit-cards.pdf
[the remark this answers is not in the document]
Comment (CodeChamp-SS) [reply gemini/sample-apps/customer-search/files/CymbalBankWebDeployed/public/cymbal-credit-cards.pdf:1]: Missed this file, please add this one too in the GCS bucket.

File: gemini/sample-apps/customer-search/files/CymbalBankWebDeployed/style/style_new.css
[the remark this answers is not in the document]
Comment (CodeChamp-SS) [reply gemini/sample-apps/customer-search/files/CymbalBankWebDeployed/style/style_new.css:1]: Initially we had two versions of css styling, removed the older one and renamed the file

File: gemini/sample-apps/customer-search/files/tnc.jpeg
[the remark this answers is not in the document]
Comment (CodeChamp-SS) [reply gemini/sample-apps/customer-search/files/tnc.jpeg:1]: This is no longer used, deleted it

File: gemini/sample-apps/customer-search/functions/account-health-summarisation-optimised/main.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+        if row["avg_monthly_balance"] is not None:
+            avg_monthly_balance = int(row["avg_monthly_balance"])
+
+    for row in res["query_assets"]:
+        if row["asset"] is not None:
+            asset_amount = int(row["asset"])
+
+    for row in res["query_debts"]:
+        if row["debt"] is not None:
+            debt_amount = int(row["debt"])
```
[the remark this answers is not in the document]
Comment (CodeChamp-SS) [reply gemini/sample-apps/customer-search/functions/account-health-summarisation-optimised/main.py]: Have created a function now to get these values

File: gemini/sample-apps/customer-search/functions/account-health-summarisation-optimised/main.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    print("Account Status = ", account_status)
+
+    vertexai.init(project=project_id, location="us-central1")
+    parameters = {
+        "max_output_tokens": 512,
+        "temperature": 0.2,
+        "top_p": 0.8,
+        "top_k": 40,
+    }
+    model = TextGenerationModel.from_pretrained("text-bison")
```
[the remark this answers is not in the document]
Comment (CodeChamp-SS) [reply gemini/sample-apps/customer-search/functions/account-health-summarisation-optimised/main.py]: Have replaced text-bison with Gemini everywhere.

File: gemini/sample-apps/customer-search/functions/event-recommendationv2/main.py
Comment (CodeChamp-SS) [reply gemini/sample-apps/customer-search/functions/event-recommendationv2/main.py:1]: Renamed the functions to more meaningful names.

File: gemini/sample-apps/customer-search/functions/create_fd/main.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    Args:
+      dob: A datetime.date object representing the person's date of birth.
+
+    Returns:
+      True if the person is a senior citizen, False otherwise.
+    """
+
+    today = datetime.date.today()
+    age = today.year - dob.year - ((today.month, today.day) < (dob.month, dob.day))
+    print("Today = ", today)
```
[the remark this answers is not in the document]
Comment (CodeChamp-SS) [reply gemini/sample-apps/customer-search/functions/create_fd/main.py]: Have removed unnecessary print statements from all the files.

File: gemini/sample-apps/customer-search/files/init_bqml_and_vector_search.ipynb
[the remark this answers is not in the document]
Comment (CodeChamp-SS) [reply gemini/sample-apps/customer-search/files/init_bqml_and_vector_search.ipynb:1]: moved them in the top directory

**Comment by CodeChamp-SS** [comment]
Hi @holtskinner, we have tried to resolve the suggestions you made, the other lint errors. I have created a new pull request #700 , please review it again. Thanks!

**Comment by CodeChamp-SS** [comment]
> Ok, you didn't need to make a new Pull Request for the changes, but it's okay.
> 
> Please close out this Pull Request if it's no longer needed.

this branch got a bit messy, so I thought creating a new one would be better.

**Pull request closed without merging**